synapse/storage/data_stores/main/events_worker.py

```diff
@@ -1362,6 +1362,62 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    async def get_unread_message_count_for_user(
```

Comment:
I think this is going to need a cache for efficiency, and I think the cache will work best if you bring the logic for populating `last_read_event_id` inside the function.

of course, that means we'll have to worry about cache invalidation...